src/proxy_lib/proxy_lib.c

```diff
@@ -0,0 +1,263 @@
+    }
+    // The UMF pool has just been created (Proxy_pool != NULL). Stop using
+    // the linear allocator and start using the UMF pool allocator from now on.
+}
+
+void proxy_lib_destroy_common(void) {
+    umf_memory_pool_handle_t pool = Proxy_pool;
+    Proxy_pool = NULL;
+    umfPoolDestroy(pool);
+    umfMemoryProviderDestroy(OS_memory_provider);
+    // the "LEAK" linear base allocator is not destroyed at all
+}
+
+/*****************************************************************************/
+/*** Generic version of realloc() of linear base allocator *******************/
+/*****************************************************************************/
+
+static inline void *ba_generic_realloc(umf_ba_linear_pool_t *pool, void *ptr,
+                                       size_t size) {
+    if (size == 0) {
+        // it means free(ptr), but linear base allocator does not implement free()
+        return NULL;
+    }
+
+    assert(ptr); // it is verified in the main realloc()
+
+    size_t max_size = umf_ba_linear_pool_contains_pointer(pool, ptr);
+    assert(max_size > 0); // assert that pool contains the pointer
+
+    void *new_ptr = umf_ba_linear_alloc(pool, size);
+    if (!new_ptr) {
+        return NULL;
+    }
+
+    if (size > max_size) {
+        size = max_size;
+    }
+
+    memcpy(new_ptr, ptr, size);
+
+    return new_ptr;
+}
+
+/*****************************************************************************/
+/*** The "LEAK" linear base allocator functions ******************************/
+/*****************************************************************************/
+
+static void ba_leak_create(void) { Base_alloc_leak = umf_ba_linear_create(0); }
+
+// it does not implement destroy(), because it will not free memory at all
+
+static inline void *ba_leak_malloc(size_t size) {
+    util_init_once(&Base_alloc_leak_initialized, ba_leak_create);
+    return umf_ba_linear_alloc(Base_alloc_leak, size);
+}
+
+static inline void *ba_leak_calloc(size_t nmemb, size_t size) {
+    util_init_once(&Base_alloc_leak_initialized, ba_leak_create);
+    // umf_ba_linear_alloc() returns zeroed memory
+    return umf_ba_linear_alloc(Base_alloc_leak, nmemb * size);
+}
+
+static inline void *ba_leak_realloc(void *ptr, size_t size) {
+    util_init_once(&Base_alloc_leak_initialized, ba_leak_create);
+    return ba_generic_realloc(Base_alloc_leak, ptr, size);
+}
+
+static inline void *ba_leak_aligned_alloc(size_t alignment, size_t size) {
+    util_init_once(&Base_alloc_leak_initialized, ba_leak_create);
+    void *ptr = umf_ba_linear_alloc(Base_alloc_leak, size + alignment);
+    return (void *)ALIGN_UP((uintptr_t)ptr, alignment);
+}
+
+static inline size_t ba_leak_pool_contains_pointer(void *ptr) {
+    return umf_ba_linear_pool_contains_pointer(Base_alloc_leak, ptr);
+}
+
+/*****************************************************************************/
+/*** The UMF pool allocator functions (the public API) ***********************/
```

Comment:
Should we support other functions, like posix_memaling first? Perhaps we should look at what's supported in https://github.com/pmem/vmem/blob/master/src/libvmmalloc/libvmmalloc.c

Reply:
The UMF pool API (`umf_memory_pool_ops_t`) supports only the following ones:
```c
    void *(*malloc)(void *pool, size_t size);
    void *(*calloc)(void *pool, size_t num, size_t size);
    void *(*realloc)(void *pool, void *ptr, size_t size);
    void *(*aligned_malloc)(void *pool, size_t size, size_t alignment);
    size_t (*malloc_usable_size)(void *pool, void *ptr);
    umf_result_t (*free)(void *pool, void *);
```

Reply:
We can add more functions when we all agree what API exactly should be supported. It can be added in a separate PR too.